base/distributed/messages.jl

```diff
@@ -80,34 +80,41 @@ end
 # of approximately 10%. Can be removed once module Serializer
 # has been suitably improved.
 
-# replace  CallMsg{Mode} with specific invocations
-const msgtypes = filter!(x->x!=CallMsg, subtypes(AbstractMsg))
-push!(msgtypes, CallMsg{:call}, CallMsg{:call_fetch})
+const msgtypes = Any[CallWaitMsg, IdentifySocketAckMsg, IdentifySocketMsg,
+                     JoinCompleteMsg, JoinPGRPMsg, RemoteDoMsg, ResultMsg,
+                     CallMsg{:call}, CallMsg{:call_fetch}]
 
 for (idx, tname) in enumerate(msgtypes)
-    nflds = length(fieldnames(tname))
-    @eval begin
-        function serialize(s::AbstractSerializer, o::$tname)
-            write(s.io, UInt8($idx))
-            for fld in fieldnames($tname)
-                serialize(s, getfield(o, fld))
-            end
-        end
-
-        function deserialize_msg(s::AbstractSerializer, ::Type{$tname})
-            data=Array{Any,1}($nflds)
-            for i in 1:$nflds
-                data[i] = deserialize(s)
-            end
-            return $tname(data...)
-        end
+    exprs = Any[ :(serialize(s, o.$fld)) for fld in fieldnames(tname) ]
+    @eval function serialize_msg(s::AbstractSerializer, o::$tname)
+        write(s.io, UInt8($idx))
+        $(exprs...)
+        return nothing
     end
 end
 
 function deserialize_msg(s::AbstractSerializer)
     idx = read(s.io, UInt8)
-    t = msgtypes[idx]
-    return eval(current_module(), Expr(:body, Expr(:return, Expr(:call, deserialize_msg, QuoteNode(s), QuoteNode(t)))))
+    if idx == 1
+        return CallWaitMsg(deserialize(s), deserialize(s), deserialize(s))
+    elseif idx == 2
+        return IdentifySocketAckMsg()
+    elseif idx == 3
+        return IdentifySocketMsg(deserialize(s))
+    elseif idx == 4
+        return JoinCompleteMsg(deserialize(s), deserialize(s))
+    elseif idx == 5
+        return JoinPGRPMsg(deserialize(s), deserialize(s), deserialize(s), deserialize(s))
+    elseif idx == 6
+        return RemoteDoMsg(deserialize(s), deserialize(s), deserialize(s))
+    elseif idx == 7
+        return ResultMsg(deserialize(s))
+    elseif idx == 8
+        return CallMsg{:call}(deserialize(s), deserialize(s), deserialize(s))
+    elseif idx == 9
+        return CallMsg{:call_fetch}(deserialize(s), deserialize(s), deserialize(s))
+    end
+    assert(false)
```

Comment:
I could generate this code automatically, but since there are only 9 cases I wasn't sure it would actually be easier to read. I can try it that way.